### DISCUSSION TITLE
Release 1.20

### DISCUSSION THREAD

**Discussion by webspiderteam** [webspiderteam/Bluetooth-DMM-For-Windows#25] [Announcements]
- Fixed Resize window when chart off
- Changed TrayIcon menu 
- Performance improvement (while not connected Cpu Usege was about 6-9%)
- Improve Popup Windows Move action stability
- Reduce Cpu usage while waiting for Connection window response
- Fixed DataChart not shown when activated from Settings

<hr /><em>This discussion was created from the release <a href='https://github.com/webspiderteam/Bluetooth-DMM-For-Windows/releases/tag/WPFRelease1.20'>Release 1.20</a>.</em>

**Comment by ludwich66** [comment]
Did you change Difference Color between Active an background Window more?
It is very good with this shadow view.
+++
L

**Reply by webspiderteam** [reply]
No i didnt changed the color just added blur effect to in active window

**Reply by webspiderteam** [reply]
what happened? I saw that you say there were a issue

**Reply by ludwich66** [reply]
i had problems to start App, the startet, minimized an closed.
I found the AppDate Webspider Folder and killed it. Now it works.
Did you activated Raw-Data and Minimize by default?
L

**Reply by webspiderteam** [reply]
both of them default enabled until I added them not changed default values

**Reply by webspiderteam** [reply]
> i had problems to start App, the startet, minimized an closed.

Probably file corrupted for some reason.

**Comment by ludwich66** [comment]
MQTT Secure with Password (not Tested with Certificate) works.
This was a hard birth, Mosquito changed secure settings from last Installation to actual.
Test from MQTT works on localhost, but first after changing configfile you can use it from other PC in the Net.
learning by curse ;-(
L

**Reply by webspiderteam** [reply]
Certificate? Is there a certificate and for what?

**Reply by ludwich66** [reply]
you can use TLS transport and bind it on a CA. 
But i dont know - i use it only local not for external locations in WAN.

https://mosquitto.org/man/mosquitto-tls-7.html

**Comment by ludwich66** [comment]
Stability Issue,

Hi Engin,
i made with 1.20 two longtime tests. measurement was two times canceled (over night).
App was closed, started then without problems.
PC and Broker are Running without deep sleep etc.

I measured temperature but only wrote changes to file. So I can not say how many measured values were generated before the termination.

Any ideas how we can narrow this down?

- Maximum number of readings reached?

Have a nice Day!
Ludwich

**Reply by webspiderteam** [reply]
> * Maximum number of readings reached?

Not posible teorically its 1M seconds. also I did measuremente about 10 hours without problem. there shoul be unexpectable data and it has to write to log if there is. Also bluetooth adapter can be issue sometimes bluetooth losts connecttion and even resets it selfs. I didnt tested bluetooth adapter issues.

**Reply by webspiderteam** [reply]
I tried to simulate Bluetooth and Mqtt server loss but no issue. but I couldnt try bluetooth compeletely because mine is integrated and cant pull it out just disabled. Maybe you can try

**Reply by ludwich66** [reply]
How do i write a debug log - is there an Option?
L

**Reply by webspiderteam** [reply]
No option for it ı can only see them in visual studio there is no option for terminated so it can't write actual problem. Did you found when it's crashing

**Comment by ludwich66** [comment]
It runs, i save MQTT Datepackages now complete, collect 300 Messages and save it on block.

lets Look ;-)

L

**Reply by webspiderteam** [reply]
I think I found the problem. Its mqtt. you installed new broker right? so whats the protocol version? I tried a new mqtt server and ım getting errors.

**Reply by ludwich66** [reply]
I dont know what Protocoll Version is negotiated

MQTT Support:
Mosquitto supports MQTT v5.0, v3.1.1, and v3.1.

My Progrannversion: mosquitto version 2.0.11 starting

NodeRed View:
Traditionel 3.1 ist not activated 
<img width="358" alt="image" src="https://user-images.githubusercontent.com/12202733/186680906-f2d23817-e1c3-4f5e-964e-a2c5852b0fea.png">

--> Then 3.1.1 or 5.0 is working

Test measurement Row:
It works from yesterday 08:45 to 13:59 correct
I have seen now, does bt-dmm programm was closed.

L

**Reply by webspiderteam** [reply]
Is app crashed? If it is go to windows "Event Viewer" app. May be there is a clue... 
Im working on crash reporter now but if we can find it its better.

**Reply by ludwich66** [reply]
Please

14.52:35
Anwendung: BluetoothDMM.exe
Frameworkversion: v4.0.30319
Beschreibung: Der Prozess wurde aufgrund einer unbehandelten Ausnahme beendet.
Ausnahmeinformationen: System.OutOfMemoryException
   bei System.Threading.Thread.StartInternal(System.Security.Principal.IPrincipal, System.Threading.StackCrawlMark ByRef)
   bei System.Threading.Thread.Start(System.Threading.StackCrawlMark ByRef)
   bei System.Threading.Thread.Start()
   bei uPLibrary.Networking.M2Mqtt.Fx.StartThread(System.Threading.ThreadStart)
   bei uPLibrary.Networking.M2Mqtt.MqttClient.Connect(System.String, System.String, System.String, Boolean, Byte, Boolean, System.String, System.String, Boolean, UInt16)
   bei uPLibrary.Networking.M2Mqtt.MqttClient.Connect()
   bei uPLibrary.Networking.M2Mqtt.MqttClient.ReconnectThread()
   bei System.Threading.ThreadHelper.ThreadStart_Context(System.Object)
   bei System.Threading.ExecutionContext.RunInternal(System.Threading.ExecutionContext, System.Threading.ContextCallback, System.Object, Boolean)
   bei System.Threading.ExecutionContext.Run(System.Threading.ExecutionContext, System.Threading.ContextCallback, System.Object, Boolean)
   bei System.Threading.ExecutionContext.Run(System.Threading.ExecutionContext, System.Threading.ContextCallback, System.Object)
   bei System.Threading.ThreadHelper.ThreadStart()

14:52:36
Name der fehlerhaften Anwendung: BluetoothDMM.exe, Version: 1.20.0.0, Zeitstempel: 0x6301077f
Name des fehlerhaften Moduls: KERNELBASE.dll, Version: 10.0.19041.1889, Zeitstempel: 0xdc0d8494
Ausnahmecode: 0xe0434352
Fehleroffset: 0x0012ca42
ID des fehlerhaften Prozesses: 0x3e1c
Startzeit der fehlerhaften Anwendung: 0x01d8b78a241c1a33
Pfad der fehlerhaften Anwendung: C:\Users\reteq\Downloads\BLUETOOTH_DMM\BluetoothDMMForWindows120.Portable\BluetoothDMM.exe
Pfad des fehlerhaften Moduls: C:\Windows\System32\KERNELBASE.dll
Berichtskennung: d896bc2b-f6a2-498e-a830-c275451910e7
Vollständiger Name des fehlerhaften Pakets: 
Anwendungs-ID, die relativ zum fehlerhaften Paket ist:

**Reply by webspiderteam** [reply]
It look like happens when your Ram or cache file full. Did you check ram usage?

**Reply by ludwich66** [reply]
10,9 GB from 16 Available
with restart BT-DMM used 0,1 % CPU and 26,6 MB RAM; LAN 0,1 MBit/s

On this machine is only BT-DDM runing
Brocker is another SBC and 
MQTT with Staroage is a Third SBC.


I look in the evening again and look of an increasuing usage.

L

**Reply by webspiderteam** [reply]
Pagefile?

**Reply by ludwich66** [reply]
What did you mean with Pagefile.sys?
At the Moment 2,4 GB other than that I can't say.
L

**Reply by webspiderteam** [reply]
Too low for 16gb ram if it is 2.4gb. pagefile can cause crashes. I have 8gb ram and my page file is now 13.7 gb. İf you changed it manually or if windows installed before ram upgrade this can cause that side effect

**Reply by webspiderteam** [reply]
Typically it should be between half of ram size and twice of ram size in your condition it should be 8096-32392

**Comment by ludwich66** [comment]
good way

how do i change pagefile - i meen size is automatic increase or decrease ?

L

**Reply by webspiderteam** [reply]
Right click to my computer. Enter properties
İn there at right side there should be advanced system settings. İn there at advanced tab you will see performance settings. İn performance settings advanced tab you will see virtual memory. Click change. Can you take a screenshot of it

**Reply by webspiderteam** [reply]
![image](https://user-images.githubusercontent.com/85828505/186749345-6a8a450c-f0f2-4487-9fd2-989512527242.png)
thats my setting rightnow
if you have more than 30gb free space you can set like 
![image](https://user-images.githubusercontent.com/85828505/186749715-f13cd0d6-5ef4-4c82-8951-6b61606ae881.png)
then press "Set" before OK

**Comment by webspiderteam** [comment]
[Portable.zip](https://github.com/webspiderteam/Bluetooth-DMM-For-Windows/files/9427989/Portable.zip)
also use this. added log for crashes

**Reply by ludwich66** [reply]
Engin Thanks,

good manuel. I changed it and try another longtime maesurement.
i change Programm version after the next chrash - i hope it runs now without problems.

Cu tomorrow !!

L

+++

**Comment by ludwich66** [comment]
Good morning Engin,

Tomorrow the same App shutdown, (with 1.20 original).
Errors in Log are the same as before.

Now i changed to your debugversion.
Start up with:

![image](https://user-images.githubusercontent.com/12202733/186861595-e8ab371f-1697-4154-8500-030f84fe0d26.png)

and after short time another window, this shut down App.

L

**Reply by webspiderteam** [reply]
can you send me crash log

**Reply by webspiderteam** [reply]
under app dir

**Comment by ludwich66** [comment]
Sorry, here a sample of all Crashlogfiles from today

[crashlog_26-08-2022_10_35_27.zip](https://github.com/webspiderteam/Bluetooth-DMM-For-Windows/files/9432436/crashlog_26-08-2022_10_35_27.zip)

L

**Reply by webspiderteam** [reply]
All errors are depend on mqtt library. it loses connection and then gives error. is your server localhost? or remote device?
anyway I should change mqtt library. but most of them has bigger files.

**Reply by ludwich66** [reply]
Engin, here my topology

ANENG V05B -> BluetoothLE
(MQTT Client) Windows 10 PC (MQTT Client) BT-DMM 1.20
(MQTT BROKER) OrangepiZeroPlus LINUX 5.15 , MOSQUITTO V. 2.0.11 (OPi with 512 MB RAM / 8GB eMMC only this Mainjob) 
(MQTT Client) NodeRed on a ODROID 4GB RAM

L

**Reply by webspiderteam** [reply]
Posibly there is connection loss sometimes an mqtt client cant handle it right way

**Comment by ludwich66** [comment]
Engin,
short Question:
Why does your debug version find problems with start up?
The App does run over 24 Hours without crashes or data loss?

**Reply by webspiderteam** [reply]
there is no error in my pc.. also all errors are depend on mqtt wifi connections are not so stable and SoC are also much more unstable wifi

**Comment by webspiderteam** [comment]
Can you try this
[Portable.zip](https://github.com/webspiderteam/Bluetooth-DMM-For-Windows/files/9435492/Portable.zip)

**Reply by ludwich66** [reply]
Yes - what did it?

**Reply by webspiderteam** [reply]
changed mqtt client library

**Reply by ludwich66** [reply]
It Runs
Do you need Responses?
I spent another night ;-)
L

**Reply by webspiderteam** [reply]
I need to know if issue fixed but no hurry for it

**Reply by ludwich66** [reply]
time will tell
it runs, i went to FamilyTV evening now
will be back at 22:00

**Reply by webspiderteam** [reply]
relax Ludwich let it run I can wait.

**Reply by ludwich66** [reply]
Morning Engin,

Good News its run-in and running...

47734 Measurements
62 Messages  false, with losses and double received values

Its around 1/1000 Promile - it's Ok. (for me)

Doubles or Transmition-errors:

Did your new MQTT Client handle QOS Levels?
I used completely QOS 0 

QOS 0  Minimal traffic
<img width="649" alt="image" src="https://user-images.githubusercontent.com/12202733/187024868-29c3bf20-e74c-4f53-87fb-5fbcbd3cce9a.png">

QOS 1 Double Traffic
<img width="640" alt="image" src="https://user-images.githubusercontent.com/12202733/187024798-a885fd3f-c1fa-4e21-ad04-d9d8aae9267c.png">

QOS 2 Quadruple Traffic
<img width="640" alt="image" src="https://user-images.githubusercontent.com/12202733/187024849-9745a4ae-18bb-4452-884b-70c9978e14bd.png">

And this pro Clientside

I have no experience of the high level "burden". But we can test that.

Q: https://www.youtube.com/watch?v=QhecPNkWBQ0
EDI's Techlab A very compact clip about MQTT with all relevant Infos
Its in German(Austria) but you can choose undertitle.

L

**Reply by webspiderteam** [reply]
> 62 Messages false, with losses and double received values

you mean loose probes connection? if displays goes auto or no value shown double value  error can happen. I cant do anything for it. BTW are there error window while these errors or just seen on the log? if it gives errors with error window I have to remove the window.
if there is debub log can you send me?


> Did your new MQTT Client handle QOS Levels?

I dont know but trobably. both clients can get and send datas. and this one is most known Library for Mqtt

**Comment by ludwich66** [comment]
I mean Datapackages from Source to Sink (CLIENT:V05B/BTDMM ->  BROKER -> NodeRed:CLIENT)

Did you use this MQTT LiB?
https://github.com/dotnet/MQTTnet

there is this comment:
`Also included ManagedMqttClient which maintains the connection and subscriptions automatically. Also application messages are queued and re-scheduled for higher QoS levels automatically.`

I must start another Testrow later -- Battery Low Indicator works ;-)
It works for 3 Days with 900 mAh NiMh cells. V05B uses smaller batteries then AN9002
Its al lot of Time for continuous run.

I have Bat-Low not in my NodeRed Sample implemented now. if time ... 

L

**Reply by webspiderteam** [reply]
> Did you use this MQTT LiB?
> https://github.com/dotnet/MQTTnet

yes thats it.

> It works for 3 Days with 900 mAh NiMh cells.

its short for me. I didnt changed my batteries until I buy it. Is about 2 years. Thats why I chose this long battery life more accurated than my old. an it was using 9v batteries and I had to change every mount. also I was forgetting to shut it of. 9v batteries not good also. and also bluetooth ability of course.

**Comment by webspiderteam** [comment]
If its ok I'm releasing this.

**Reply by ludwich66** [reply]
Looks good!

- No App Crashes.
- 135.600 readings 

;-) +++

L

**Reply by webspiderteam** [reply]
Almost 40 hours 😊

**Comment by ludwich66** [comment]
You are welcome (if you are bored) to install a normal counter for the MQTT string to select from.
It is then easier to find "mistransmissions".
I have to parse the time in Libre Calc first ;-)
L

**Comment by ludwich66** [comment]
I'll then switch to the official 1.21, new batteries and mega long range test ;-)

**Reply by webspiderteam** [reply]
I hope this one is last test. I think thats enough for this app.

**Reply by webspiderteam** [reply]
I also released Mqtt test app with server ability.

**Reply by ludwich66** [reply]
I am more than satisfied with the result !!
